deployment_logger.go

```diff
@@ -0,0 +1,254 @@
+// Copyright 2016 Mender Software AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+package main
+
+import (
+	"bufio"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"sort"
+	"strconv"
+)
+
+// error messages
+var (
+	ErrNoLogFiles           = errors.New("log directory empty")
+	ErrLoggerNotInitialized = errors.New("logger not initialized")
+	ErrNoDeploymentId       = errors.New("no valid deployment id")
+	ErrParsingLogFile       = errors.New("can not parse deployment log file")
+)
+
+// JSONFileLogger struct
+type FileLogger struct {
+	logFileName string
+	logFile     io.WriteCloser
+}
+
+// NewJSONFileLogger creates instance of file logger; it is initialized
+// just before logging is started
+func NewFileLogger(name string) *FileLogger {
+	// open log file
+	logFile, err := os.OpenFile(name, os.O_RDWR|os.O_CREATE|os.O_APPEND, 0666)
+	if err != nil {
+		// if we can not open file for logging; return nil
+		return nil
+	}
+
+	// return FileLogger only when logging is possible (we can open log file)
+	return &FileLogger{
+		logFileName: name,
+		logFile:     logFile,
+	}
+}
+
+func (fl *FileLogger) Write(log []byte) (int, error) {
+	return fl.logFile.Write(log)
+}
+
+func (fl *FileLogger) Deinit() error {
+	return fl.logFile.Close()
+}
+
+type DeploymentLogManager struct {
+	logLocation         string
+	logFileName         string
+	logDeploymentIdFile string
+	deploymentStore     *DirStore
+
+	deploymentID string
+	logger       *FileLogger
+	// how many log files we are keeping in log directory before rotating
+	maxLogFiles int
+	// it is easy to add logging hook, but not so much remove it;
+	// we need a mechanism for emabling and disabling logging
+	loggingEnabled bool
+}
+
+func NewDeploymentLogManager(logDirLocation string) *DeploymentLogManager {
+	return &DeploymentLogManager{
+		logLocation: filepath.Dir(logDirLocation),
+		logFileName: filepath.Base(logDirLocation),
```

Comment:
Maybe the user should not be able to change the log file name, because it's an internal part of the client. Instead I'd try to enforce a naming scheme that makes it easy to extract deployment ID, for instance: `"%04d.<deployment-id>"`. Log ID would come first, then deployment ID (maybe with some escaping so that it can be used as part of a filename).

Also, log location should perhaps be relative to the global data path location, that defaults to `/var/lib/mender`, but can be overridden by passing `-data` in command line.